Add REST API endpoints for analytics batch import status and manual triggering

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR implements REST API endpoints for analytics batch import status and manual triggering, as part of the Core Analytics Scheduled Imports UI feature (WOOPLUG-5761).

Closes WOOA7S-793 

#### What was added:

1. **New REST API Controller** (`src/Admin/API/AnalyticsImports.php`)
   - Only loads when both `analytics` and `analytics-scheduled-import` features are enabled
   - Extends `WC_REST_Data_Controller` following existing patterns

2. **GET `/wc-analytics/imports/status` Endpoint**
   - Returns current import mode (`immediate` or `scheduled`)
   - In scheduled mode, provides:
     - `last_processed_date` - Last processed order date in site timezone
     - `next_scheduled` - Next scheduled import time in site timezone  
     - `import_in_progress_or_due` - Boolean indicating if manual import is already in progress or due to run
   - In immediate mode, scheduled-related fields return `null`
   - Requires `manage_woocommerce` capability

3. **POST `/wc-analytics/imports/trigger` Endpoint**
   - Manually triggers batch import in scheduled mode
   - Returns 400 error if in immediate mode
   - Returns 400 error if manual import already scheduled
   - Requires `manage_woocommerce` capability

4. **Comprehensive Unit Tests** (`tests/php/src/Admin/API/AnalyticsImportsTest.php`)
   - 11 tests covering all functionality
   - Tests for both endpoints in different modes
   - Permission checks for admin, shop manager, and customer roles
   - Timezone conversion validation

#### Key Implementation Details:

- Datetimes are converted from GMT to site timezone using `get_date_from_gmt()`
- Manual import detection checks for specifically triggered imports (not recurring scheduled actions)
- Controller registration is conditional on feature flags

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions):

#### Prerequisites:

1. Install the WooCommerce Beta Tester plugin
2. Go to `/wp-admin/tools.php?page=woocommerce-admin-test-helper&tab=features` and enable the `analytics-scheduled-import` feature flag

#### Test Status Endpoint (Immediate Mode):

1. Set import mode to immediate: `wp option update woocommerce_analytics_immediate_import yes`
2. Make a GET request to `/wp-json/wc-analytics/imports/status`

```js
wp.apiFetch( {
		path: '/wc-analytics/imports/status',
		method: 'GET',
}).then( ( response ) => {
	console.log( response );
}).catch( ( error ) => {
	console.error( error );
});
```

3. Verify response contains:
   - `mode: "immediate"`
   - All other fields should be `null`


#### Test Status Endpoint (Scheduled Mode):

1. Set import mode to scheduled: `wp option update woocommerce_analytics_immediate_import no`
2. Wait for ~10 seconds for the next scheduled action to be set
3. Make a GET request to `/wp-json/wc-analytics/imports/status`
4. Verify response contains:
   - `mode: "scheduled"`
   - `last_processed_date` (should be a datetime string in site timezone or null)
   - `next_scheduled` (should be a datetime string in site timezone or null)
   - `import_in_progress_or_due` (should be `false`)
5. Go to `/wp-admin/admin.php?page=wc-status&tab=action-scheduler&status=pending` and search for `wc-admin_process_pending_orders_batch` and verify the scheduled time (in GMT) matches the `next_scheduled` field.

<img width="1146" height="295" alt="Screenshot 2025-12-02 at 13 56 18" src="https://github.com/user-attachments/assets/95ca6e68-88a3-4699-a62f-e68f700539bf" />

<img width="1147" height="596" alt="Screenshot 2025-12-02 at 13 55 49" src="https://github.com/user-attachments/assets/7804a643-15df-4638-a1a2-060d3fb3afa9" />

#### Test Manual Trigger Endpoint:

1. Set import mode to scheduled: `wp option update woocommerce_analytics_immediate_import no`
2. Make a POST request to `/wp-json/wc-analytics/imports/trigger`

```js
wp.apiFetch( {
		path: '/wc-analytics/imports/trigger',
		method: 'POST',
}).then( ( response ) => {
	console.log( response );
}).catch( ( error ) => {
	console.error( error );
});
```

3. Verify response shows `success: true` with success message
4. Make another POST request immediately
5. Verify it returns 400 error indicating import already scheduled

#### Test Immediate Mode Restriction:

1. Set import mode to immediate: `wp option update woocommerce_analytics_immediate_import yes`
2. Make a POST request to `/wp-json/wc-analytics/imports/trigger`
3. Verify it returns 400 error with message about immediate mode

#### (Optional) Test Permissions:

1. As a customer user, make GET/POST requests to both endpoints
2. Verify both return 403 Forbidden errors
3. As shop manager, verify both endpoints are accessible

### Testing that has already taken place:

- All 11 PHPUnit tests passing (11 tests, 37 assertions)
- Tests cover:
  - Status endpoint in both immediate and scheduled modes
  - Manual trigger success and error cases  
  - Permission checks for different user roles
  - Datetime timezone conversion
  - Manual import detection

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

chihsuan committed:
Add REST API Analytics Imports Controller for batch import management

This commit introduces a new class, AnalyticsImports, which handles requests for checking batch import status and triggering manual imports. It includes permission checks for user access and defines the necessary REST API routes for these functionalities. Additionally, the Init class is updated to register the new controller if the 'analytics-scheduled-import' feature is enabled.

chihsuan committed:
Add unit tests for AnalyticsImports API controller

This commit introduces a new test class, AnalyticsImportsTest, which includes comprehensive unit tests for the Analytics Imports API controller. The tests cover various scenarios such as checking the status of imports, triggering imports, and verifying user permissions. This addition ensures the functionality and reliability of the API endpoints related to batch import management.

chihsuan committed:
Add changelog

chihsuan committed:
Update plugins/woocommerce/tests/php/src/Admin/API/AnalyticsImportsTest.php

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

chihsuan committed:
Fix lint

chihsuan committed:
Clean up options

kangzj commented on plugins/woocommerce/src/Admin/API/AnalyticsImports.php:
```diff
@@ -0,0 +1,279 @@
+	 * @return \WP_Error|boolean
+	 */
+	public function create_item_permissions_check( $request ) {
+		if ( ! current_user_can( 'manage_woocommerce' ) ) {
```
This looks the same as `get_item_permissions_check`. Is there another capability to check? Otherwise, could we merge the two?

chihsuan replied:
Yeah, this is the same as `get_item_permissions_check`. Sounds good. I'll merge them.

chihsuan replied:
Done in 5c0410644e1d5eac5557fc8f36ae977fd009d9d5

kangzj commented on plugins/woocommerce/src/Admin/API/AnalyticsImports.php:
```diff
@@ -0,0 +1,279 @@
+		}
+
+		// Schedule the batch import action.
+		OrdersScheduler::schedule_action( 'process_pending_batch', array( null, null ) );
```
I wonder if we should use a more distinct name for the action as the name looks very general to me, which might be less expressive and easier to cause conflict? Also maybe we could use constant?

chihsuan replied:
This name is defined in `OrdersScheduler.php,` and it's more like an alias for the real action name. The real action name that is used for scheduling is `wc-admin_process_pending_orders_batch`, so it shouldn't cause any conflict.

https://github.com/woocommerce/woocommerce/blob/847e0168140b898e0b350b9c60c680e32338c301/plugins/woocommerce/src/Internal/Admin/Schedulers/OrdersScheduler.php#L119-L133

kangzj commented on plugins/woocommerce/src/Admin/API/AnalyticsImports.php:
```diff
@@ -0,0 +1,279 @@
+	 * @return bool True if a manual triggered import has already been scheduled, false otherwise.
+	 */
+	private function has_manual_triggered_import_scheduled() {
+		$has_scheduled_action = function_exists( 'as_has_scheduled_action' ) ? 'as_has_scheduled_action' : 'as_next_scheduled_action';
```
I wonder where this function lives and whether `as_next_scheduled_action` will always exist.

chihsuan replied:
This function is defined in the [action-scheduler package](https://actionscheduler.org/). Old versions of the package may not have this function. I used the same pattern as in the existing code to check if the function exists before using it. https://github.com/woocommerce/woocommerce/blob/847e0168140b898e0b350b9c60c680e32338c301/plugins/woocommerce/src/Blocks/BlockPatterns.php#L209-L211

However, that's a valid observation. Since the Action Scheduler package is bundled with WooCommerce, there's no need to verify the existence of the action before invoking it. I'll make the necessary adjustments.

chihsuan replied:
Simplified it in af6c46f3e82cc365ff459a7c1c538e911feba075

kangzj commented on plugins/woocommerce/src/Admin/API/AnalyticsImports.php:
```diff
@@ -0,0 +1,279 @@
+		}
+
+		// Determine if a batch import has already been scheduled or is in progress, excluding the recurring import action.
+		if ( $this->has_manual_triggered_import_scheduled() ) {
```
I ended up have two scheduled actions after manually triggering it from the endpoint. I was thinking the check would short circuit the endpoint, but it didn't catch it, it seems:

<img width="774" height="479" alt="Image" src="https://github.com/user-attachments/assets/6c332c35-26e0-435c-8291-4530b7872ff2" />

chihsuan replied:
This behavior is actually expected. Manually triggering the action creates one instance, and the recurring schedule creates another. However, this can be a bit confusing. It might be possible to trigger the recurring action immediately instead of scheduling an additional one. I'll investigate this approach to see if it offers a clearer solution.

chihsuan replied:
I've updated the PR to ensure that only one action is scheduled at a time. When manually triggering the api call, the recurring action will be canceled first and then rescheduled to run immediately.

kangzj reviewed (approved):
It works as described, and code looks good. 

[One question](https://github.com/woocommerce/woocommerce/pull/62196/files#r2591059483) asked, but don't think it's a blocker.

chihsuan committed:
Fix types

chihsuan committed:
Refactor has_manual_triggered_import_scheduled method to simplify scheduled action check

chihsuan committed:
Refactor AnalyticsImports API to improve import status checks and update response structure

- Replaced manual_triggered_import_scheduled with import_in_progress_or_due in response.
- Updated methods to check for ongoing or soon-to-run imports.
- Adjusted test cases to reflect changes in response structure and import status logic.

chihsuan committed:
Fix tests

chihsuan committed:
Fix lint

chihsuan committed:
Refactor permission checks in AnalyticsImports API controller to unify access control for analytics imports

chihsuan committed:
Fix schema

chihsuan committed:
Fix lint

chihsuan committed:
Fix phan errors

chihsuan committed:
Fix tests
